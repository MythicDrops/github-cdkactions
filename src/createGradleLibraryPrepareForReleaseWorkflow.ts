import { Job, Stack, Workflow } from "cdkactions";
import dedent from "ts-dedent";
import { GradleLibraryConfig } from "./types";

export const createGradleLibraryPrepareForReleaseWorkflow = (
  stack: Stack,
  config?: GradleLibraryConfig
): Workflow => {
  const workflow = new Workflow(stack, "prepare-for-release", {
    name: "Prepare For Release",
    on: {
      push: { branches: ["main"] },
    },
  });

  const preTestSteps = config?.preTestSteps ?? [];
  const timeoutMinutes = config?.timeoutMinutes ?? 10;

  new Job(workflow, "prepare-for-release", {
    name: "Prepare For Release",
    runsOn: "ubuntu-latest",
    timeoutMinutes,
    steps: [
      {
        name: "Check-out Code",
        uses: "actions/checkout@v2",
        with: { "fetch-depth": 0 },
      },
      {
        name: "Turnstyle",
        uses: "softprops/turnstyle@v1",
        env: {
          GITHUB_TOKEN: `\${{ secrets.GITHUB_TOKEN }}`,
        },
      },
      {
        name: "Install GitVersion",
        uses: "gittools/actions/gitversion/setup@v0.9.9",
        with: { versionSpec: "5.x" },
      },
      {
        name: "Determine Version",
        id: "gitversion",
        uses: "gittools/actions/gitversion/execute@v0.9.9",
      },
      {
        name: "Set up JDK",
        uses: "actions/setup-java@v2",
        with: {
          distribution: "adopt",
          "java-version": "8",
        },
      },
      {
        name: "Make Gradle Wrapper Executable",
        run: dedent`chmod +x ./gradlew`,
      },
      {
        name: "Set Gradle Version",
        run: dedent`./gradlew wrapper --gradle-version 7.0.2`,
      },
      ...preTestSteps, // expand preTestSteps
      {
        name: "Test Library with Gradle",
        run: dedent`./gradlew check`,
      },
      {
        name: "Push Tag",
        uses: "mathieudutour/github-tag-action@v5.1",
        with: {
          custom_tag: `\${{ steps.gitversion.outputs.semVer }}`,
          github_token: `\${{ secrets.ACTIONS_PAT }}`,
          tag_prefix: "",
          create_annotated_tag: true,
        },
      },
      {
        name: "Publish Library to Maven Central",
        run: dedent`./gradlew publishToSonatype \
          -Pversion=\${{ steps.gitversion.outputs.semVer }}`,
        env: {
          ORG_GRADLE_PROJECT_signingKey: `\${{ secrets.GPG_PRIVATE_KEY }}`,
          ORG_GRADLE_PROJECT_signingPassword: `\${{ secrets.GPG_PASSWORD }}`,
          OSSRH_USERNAME: `\${{ secrets.OSSRH_USERNAME }}`,
          OSSRH_PASSWORD: `\${{ secrets.OSSRH_TOKEN }}`,
        },
      },
      {
        name: "Close and Release Repository on Maven Central",
        run: dedent`./gradlew closeAndReleaseRepository`,
        env: {
          OSSRH_USERNAME: `\${{ secrets.OSSRH_USERNAME }}`,
          OSSRH_PASSWORD: `\${{ secrets.OSSRH_TOKEN }}`,
        },
      },
    ],
  });
  return workflow;
};
