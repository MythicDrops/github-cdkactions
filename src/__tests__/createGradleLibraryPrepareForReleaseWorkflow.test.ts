import { App, Stack } from "cdkactions";
import { readFileSync } from "fs";
import { load } from "js-yaml";
import { join } from "path";
import { createGradleLibraryPrepareForReleaseWorkflow } from "../createGradleLibraryPrepareForReleaseWorkflow";

const gradleLibraryPrepareForReleaseWorkflowRaw = readFileSync(
  join(__dirname, "gradleLibraryPrepareForReleaseWorkflow.fixture.yaml"),
  "utf-8"
);
const gradleLibraryPrepareForReleaseWorkflow = load(
  gradleLibraryPrepareForReleaseWorkflowRaw
);

describe("createGradleLibraryPrepareForReleaseWorkflow", () => {
  let app: App;
  let stack: Stack;

  beforeEach(() => {
    app = new App();
    stack = new Stack(app, "cdk");
  });

  it("generates a GitHub Actions workflow", () => {
    expect(
      createGradleLibraryPrepareForReleaseWorkflow(stack).toGHAction()
    ).toEqual(gradleLibraryPrepareForReleaseWorkflow);
  });
});
