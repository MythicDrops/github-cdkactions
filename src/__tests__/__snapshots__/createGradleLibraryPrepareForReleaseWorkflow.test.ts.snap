// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`createGradleLibraryPrepareForReleaseWorkflow generates a GitHub Actions workflow 1`] = `
Object {
  "jobs": Object {
    "prepare-for-release": Object {
      "name": "Prepare For Release",
      "runs-on": "ubuntu-latest",
      "steps": Array [
        Object {
          "name": "Check-out Code",
          "uses": "actions/checkout@v2",
          "with": Object {
            "fetch-depth": 0,
          },
        },
        Object {
          "env": Object {
            "GITHUB_TOKEN": "\${{ secrets.GITHUB_TOKEN }}",
          },
          "name": "Turnstyle",
          "uses": "softprops/turnstyle@v1",
        },
        Object {
          "name": "Install GitVersion",
          "uses": "gittools/actions/gitversion/setup@v0.9.9",
          "with": Object {
            "versionSpec": "5.x",
          },
        },
        Object {
          "id": "gitversion",
          "name": "Determine Version",
          "uses": "gittools/actions/gitversion/execute@v0.9.9",
        },
        Object {
          "name": "Set up JDK",
          "uses": "actions/setup-java@v2",
          "with": Object {
            "distribution": "adopt",
            "java-version": "8",
          },
        },
        Object {
          "name": "Make Gradle Wrapper Executable",
          "run": "chmod +x ./gradlew",
        },
        Object {
          "name": "Set Gradle Version",
          "run": "./gradlew wrapper --gradle-version 7.0.2",
        },
        Object {
          "name": "Test Library with Gradle",
          "run": "./gradlew check",
        },
        Object {
          "name": "Push Tag",
          "uses": "mathieudutour/github-tag-action@v5.1",
          "with": Object {
            "create_annotated_tag": true,
            "custom_tag": "\${{ steps.gitversion.outputs.semVer }}",
            "github_token": "\${{ secrets.ACTIONS_PAT }}",
            "tag_prefix": "",
          },
        },
        Object {
          "env": Object {
            "ORG_GRADLE_PROJECT_signingKey": "\${{ secrets.GPG_PRIVATE_KEY }}",
            "ORG_GRADLE_PROJECT_signingPassword": "\${{ secrets.GPG_PASSWORD }}",
            "OSSRH_PASSWORD": "\${{ secrets.OSSRH_TOKEN }}",
            "OSSRH_USERNAME": "\${{ secrets.OSSRH_USERNAME }}",
          },
          "name": "Publish Library to Maven Central",
          "run": "./gradlew publishToSonatype           -Pversion=\${{ steps.gitversion.outputs.semVer }}",
        },
        Object {
          "env": Object {
            "OSSRH_PASSWORD": "\${{ secrets.OSSRH_TOKEN }}",
            "OSSRH_USERNAME": "\${{ secrets.OSSRH_USERNAME }}",
          },
          "name": "Close and Release Repository on Maven Central",
          "run": "./gradlew closeAndReleaseRepository",
        },
      ],
      "timeout-minutes": 10,
    },
  },
  "name": "Prepare For Release",
  "on": Object {
    "push": Object {
      "branches": Array [
        "main",
      ],
    },
  },
}
`;
